Kernel/EFIPrekernel/Arch/aarch64/Boot.cpp

```diff
@@ -6,25 +6,262 @@
 
 #include <AK/Types.h>
 
+#include <Kernel/Arch/aarch64/ASM_wrapper.h>
+#include <Kernel/Arch/aarch64/Registers.h>
+#include <Kernel/Sections.h>
+
 #include <Kernel/EFIPrekernel/Arch/Boot.h>
+#include <Kernel/EFIPrekernel/Arch/MMU.h>
+#include <Kernel/EFIPrekernel/Panic.h>
 #include <Kernel/EFIPrekernel/Runtime.h>
+#include <Kernel/EFIPrekernel/VirtualMemoryLayout.h>
 
 namespace Kernel {
 
+// This function has to fit into one page as it will be identity mapped.
+[[gnu::aligned(PAGE_SIZE)]] [[noreturn]] NEVER_INLINE static void enter_kernel_helper(FlatPtr sctlr_el1, FlatPtr kernel_entry, FlatPtr kernel_sp, FlatPtr boot_info_vaddr) asm("enter_kernel_helper");
+[[gnu::aligned(PAGE_SIZE)]] [[noreturn]] NEVER_INLINE static void enter_kernel_helper(FlatPtr sctlr_el1, FlatPtr kernel_entry, FlatPtr kernel_sp, FlatPtr boot_info_vaddr)
+{
+    register FlatPtr x0 asm("x0") = boot_info_vaddr;
+    register FlatPtr sp asm("sp") = kernel_sp;
+    asm volatile(R"(
+        // Invalidate the TLBs before enabling the MMU, as the TLBs might still contain old values.
+        tlbi vmalle1
+        dsb ish
+        isb
+
+        msr sctlr_el1, %[sctlr_el1]
+        isb
+
+        mov lr, xzr
+        mov fp, xzr
+
+        br %[kernel_entry]
```

Comment:
Shouldn't that technically also be a goto clobber?
(never used those, so not sure)

Reply:
`asm goto` and the goto clobbers are only used for C++ labels (so the ones you use `goto` for).